File: core/cloudflow-blueprint/src/main/scala/cloudflow/blueprint/StreamletConnection.scala
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cloudflow.blueprint

import org.apache.avro._

import com.typesafe.config.Config

final case class StreamletConnection(
    from: String,
    to: String,
    label: Option[String] = None,
    problems: Vector[BlueprintProblem] = Vector.empty[BlueprintProblem],
    verified: Option[VerifiedStreamletConnection] = None,
    metadata: Option[Config] = None
) {

  def verify(verifiedStreamlets: Vector[VerifiedStreamlet]): StreamletConnection = {
    val fromPathError = VerifiedPortPath(from).left.toOption
    val toPathError = VerifiedPortPath(to).left.toOption
    val patternErrors = Set(fromPathError, toPathError).flatten

    val outletResult = VerifiedOutlet.find(verifiedStreamlets, from)
    val inletResult = VerifiedInlet.find(verifiedStreamlets, to)
    val outletError = outletResult.left.toOption
    val inletError = inletResult.left.toOption

    val verifiedFromPath = outletResult.fold(_ ⇒ from, _.portPath.toString)
    val verifiedToPath = inletResult.fold(_ ⇒ to, _.portPath.toString)

    val connectionFound = for {
      verifiedOutlet ← outletResult
      verifiedInlet ← inletResult
      validConnection ← verifySchema(verifiedOutlet, verifiedInlet, label)
    } yield validConnection
    val schemaError = connectionFound.left.toOption
    val conErrors = Set(outletError, inletError, schemaError).flatten
    val verified = connectionFound.toOption
    copy(
      problems = (patternErrors ++ conErrors).toVector,
      verified = verified,
      from = verified.fold(verifiedFromPath)(_.verifiedOutlet.portPath.toString),
      to = verified.fold(verifiedToPath)(_.verifiedInlet.portPath.toString)
    )
  }

  val AvroFormat = "avro"

  /**
   * Schema resolution is done based on the information in http://avro.apache.org/docs/current/spec.html#Schema+Resolution
   */
  private def verifySchema(
      verifiedOutlet: VerifiedOutlet,
      verifiedInlet: VerifiedInlet,
      label: Option[String]
  ): Either[BlueprintProblem, VerifiedStreamletConnection] = {

    if (verifiedOutlet.schemaDescriptor.format == verifiedInlet.schemaDescriptor.format &&
      verifiedOutlet.schemaDescriptor.fingerprint == verifiedInlet.schemaDescriptor.fingerprint) {
      Right(VerifiedStreamletConnection(verifiedOutlet, verifiedInlet, label))
    } else if (verifiedOutlet.schemaDescriptor.format == AvroFormat) {
      val writerSchema = new Schema.Parser().parse(verifiedOutlet.schemaDescriptor.schema)
      val readerSchema = new Schema.Parser().parse(verifiedInlet.schemaDescriptor.schema)

      val result = SchemaCompatibility.checkReaderWriterCompatibility(readerSchema, writerSchema).getType
      if (result == SchemaCompatibility.SchemaCompatibilityType.COMPATIBLE) Right(VerifiedStreamletConnection(verifiedOutlet, verifiedInlet, label))
      else Left(IncompatibleSchema(verifiedOutlet.portPath, verifiedInlet.portPath))
    } else Left(IncompatibleSchema(verifiedOutlet.portPath, verifiedInlet.portPath))
  }
}

File: kubectl-cloudflow/verify/streamletconnection.go
package verify

import (
	"github.com/go-akka/configuration"
	"github.com/lightbend/cloudflow/kubectl-cloudflow/cloudflowavro"
	"github.com/lightbend/cloudflow/kubectl-cloudflow/util"
)

// StreamletConnection models the connections in a blueprint
type StreamletConnection struct {
	from     string
	to       string
	label    *string
	problems []BlueprintProblem
	verified *VerifiedStreamletConnection
	metadata *configuration.Config
}

const avroFormat string = "avro"

func (c StreamletConnection) verify(verifiedStreamlets []VerifiedStreamlet) StreamletConnection {
	_, fromPathError := NewVerifiedPortPath(c.from)
	_, toPathError := NewVerifiedPortPath(c.to)

	/*
		for _, vs := range verifiedStreamlets {
			fmt.Printf("%s\n", vs.name)
		}
	*/
	var patternErrors []BlueprintProblem
	var conErrors []BlueprintProblem

	if fromPathError != nil {
		patternErrors = append(patternErrors, fromPathError)
	}

	if toPathError != nil {
		patternErrors = append(patternErrors, toPathError)
	}

	outletResult, outletError := FindVerifiedOutlet(verifiedStreamlets, c.from)
	inletResult, inletError := FindVerifiedInlet(verifiedStreamlets, c.to)

	var verifiedFromPath *string
	var verifiedToPath *string

	if outletError != nil {
		verifiedFromPath = &c.from
		conErrors = append(conErrors, outletError)
	} else {
		portPath := outletResult.portPath().ToString()
		verifiedFromPath = &portPath
	}

	if inletError != nil {
		verifiedToPath = &c.to
		conErrors = append(conErrors, inletError)
	} else {
		portPath := inletResult.portPath().ToString()
		verifiedToPath = &portPath
	}

	var verConnection *VerifiedStreamletConnection
	if outletResult != nil && inletResult != nil {
		var label = ""
		if c.label == nil {
			label = ""
		} else {
			label = *(c.label)
		}
		verConnectionRes, schemaError := c.verifySchema(*outletResult, *inletResult, label)
		verConnection = verConnectionRes
		if schemaError != nil {
			conErrors = append(conErrors, schemaError)
		}
	}

	return StreamletConnection{
		problems: append(patternErrors, conErrors...),
		verified: verConnection,
		from:     *verifiedFromPath,
		to:       *verifiedToPath,
	}
}

func (c StreamletConnection) verifySchema(verifiedOutlet VerifiedOutlet, verifiedInlet VerifiedInlet, label string) (*VerifiedStreamletConnection, BlueprintProblem) {
	if verifiedOutlet.schemaDescriptor.Format == verifiedInlet.schemaDescriptor.Format &&
		verifiedOutlet.schemaDescriptor.Fingerprint == verifiedInlet.schemaDescriptor.Fingerprint {
		return &VerifiedStreamletConnection{verifiedOutlet: verifiedOutlet, verifiedInlet: verifiedInlet, label: &label}, nil
	} else if verifiedOutlet.schemaDescriptor.Format == avroFormat {
		err := cloudflowavro.CheckSchemaCompatibility(verifiedOutlet.schemaDescriptor.Schema, verifiedInlet.schemaDescriptor.Schema)

		rerr, ok := err.(*cloudflowavro.ReaderSchemaValidationError)

		if ok {
			util.LogAndExit("Inlet schema %s with fingerprint %s is not valid. Error: %s", verifiedInlet.schemaDescriptor.Name, verifiedOutlet.schemaDescriptor.Fingerprint, rerr.Error())
		}

		werr, ok := err.(*cloudflowavro.WriterSchemaValidationError)

		if ok {
			util.LogAndExit("Outlet schema %s with fingerprint %s is not valid. Error: %s", verifiedOutlet.schemaDescriptor.Name, verifiedOutlet.schemaDescriptor.Fingerprint, werr.Error())
		}

		if err != nil {
			return nil, &IncompatibleSchema{outletPortPath: verifiedOutlet.portPath(), inletPath: verifiedInlet.portPath()}
		} else {
			return &VerifiedStreamletConnection{verifiedOutlet: verifiedOutlet, verifiedInlet: verifiedInlet, label: &label}, nil
		}

	} else {
		return nil, &IncompatibleSchema{outletPortPath: verifiedOutlet.portPath(), inletPath: verifiedInlet.portPath()}
	}
}
